File: python-puka/receive_logs_topic.py
#!/usr/bin/env python
import puka
import sys

client = puka.Client("amqp://localhost/")
promise = client.connect()
client.wait(promise)


promise = client.exchange_declare(exchange='topic_logs', type='topic')
client.wait(promise)

promise = client.queue_declare(exclusive=True)
queue_name = client.wait(promise)['queue']

binding_keys = sys.argv[1:]
if not binding_keys:
    print >> sys.stderr, "Usage: %s [binding_key]..." % (sys.argv[0],)
    sys.exit(1)

for binding_key in binding_keys:
    promise = client.queue_bind(exchange='topic_logs', queue=queue_name,
                                routing_key=binding_key)
    client.wait(promise)


print ' [*] Waiting for logs. To exit press CTRL+C'

consume_promise = client.basic_consume(queue=queue_name, no_ack=True)
while True:
    msg_result = client.wait(consume_promise)
    print " [x] %r:%r" % (msg_result['routing_key'], msg_result['body'])

File: php-thesis/topics/receive_logs_topic.php
<?php

declare(strict_types=1);

use Thesis\Amqp\Client;
use Thesis\Amqp\Config;
use Thesis\Amqp\DeliveryMessage;
use function Amp\trapSignal;

require_once __DIR__.'/../vendor/autoload.php';

$client = new Client(
    new Config(
        urls: ['rabbitmq:5672'],
    ),
);

$channel = $client->channel();

$channel->exchangeDeclare(exchange: 'topics_logs', exchangeType: 'topic');
$queue = $channel->queueDeclare(exclusive: true);

$bindingKeys = array_slice($argv, 1);
if ($bindingKeys === []) {
    file_put_contents('php://stderr', "Usage: $argv[0] [binding_key]\n");
    exit(1);
}

foreach ($bindingKeys as $bindingKey) {
    $channel->queueBind(queue: $queue->name, exchange: 'topics_logs', routingKey: $bindingKey);
}

echo " [*] Waiting for logs. To exit press CTRL+C\n";

$channel->consume(
    callback: static function (DeliveryMessage $delivery): void {
        echo " [x] {$delivery->routingKey}: {$delivery->message->body} \n";
    },
    queue: $queue->name,
    noAck: true,
);

trapSignal([\SIGTERM, \SIGINT]);

$client->disconnect();
